### PR TITLE
Last stop before CEP 40

### PR DESCRIPTION
**Pull request by ppedrot**
This PR is a potpourri of kernel tweaks and cleanups that brings the implementation of CEP 40 to a stone's throw. They can be classified in two families.
- Several invariants regarding side-effects, most notably removing all the delayed universe mumbo-jumbo that now only applies to vanilla opaque definitions.
- The explicit split in `Term_typing` of the two phases of typing of opaque definitions, first the declaration itself and then the proof.

FTR, I already have a krazy implementation of CEP 40 atop of this PR that removes the opaque table from the kernel, but does not reintroduce it anywhere else. In practice, that means that opaque proofs are happily dropped after being type-checked in the kernel. Funnily enough, a tiny one-line patch makes the whole stdlib go through, which shows that indeed opaque proofs are not often observed.

I probably need to discuss with people like @ejgallego to know where the best place to put the opaque table is before proceeding further, but in the meantime the current PR already makes sense in itself so there is no harm in merging it.

### CODE REVIEW

File: kernel/safe_typing.ml [hunk 1]
```diff
@@ -809,9 +804,9 @@ let export_private_constants ~in_section ce senv =
   let senv = List.fold_left (add_constant_aux ~in_section) senv bodies in
   (ce, exported), senv
 
-let add_constant (type a) ~(side_effect : a effect_entry) ~in_section l decl senv : (Constant.t * a) * safe_environment =
+let add_constant ~in_section l decl senv : Constant.t * safe_environment =
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml]: nit: We don't really need the type when there's no gadt stuff involved

File: tactics/declare.ml [hunk 2]
```diff
@@ -269,34 +280,48 @@ let define_constant ~side_effect ~name cd =
...
 
-let declare_private_constant ?role ?(local = ImportDefaultBehavior) ~name ~kind cd =
-  let kn, eff, export = define_constant ~side_effect:EffectEntry ~name cd in
-  let () = assert (CList.is_empty export) in
+let declare_private_constant ?role ?(local = ImportDefaultBehavior) ~name ~kind de =
+  let kn, eff =
+    let de =
+      if not de.proof_entry_opaque then
+        let body, () = Future.force de.proof_entry_body in
+        let de = { de with proof_entry_body = Future.from_val (body, ()) } in
```
Comment (SkySkimmer) [inline tactics/declare.ml]: Why do this force?

Comment (ppedrot) [reply tactics/declare.ml]: Side-effects were already forced upstream, this is just making it explicit in the kernel entry.

Comment (SkySkimmer) [reply tactics/declare.ml]: I don't get it. We force in cast_proof_entry so what's the point of these 2 lines?

Comment (ppedrot) [reply tactics/declare.ml]: I think that this is due to some typing constraint in the type of the proof_entry_body field. I'll try to see if I can simplify the code nonetheless.

Comment (ppedrot) [reply tactics/declare.ml]: Yes, indeed, I could get rid of that at the cost of changing the `proof_entry` type, but that would require messing with unrelated files. Or maybe I could have some proxy type. Does this really look important to you?

File: tactics/declare.ml [hunk 3]
```diff
@@ -264,39 +287,50 @@ let define_constant ~side_effect ~name cd =
...
+        DefinitionEff (cast_proof_entry de)
```
Comment (SkySkimmer) [inline tactics/declare.ml]: My problem is that this `cast_proof_entry` starts by forcing `de.proof_entry_body` so why not do https://github.com/coq/coq/commit/33d1b80f3e06ba9a29b45aa4171645f9a4e0a8ae

Comment (ppedrot) [reply tactics/declare.ml]: If anything, I'd prefer remove the forcing out of of `cast_proof_entry` and let take it a value, as its other caller unconditionally forces the entry. It would require less juggling with entry forcing-then-reconstruction.

Comment (SkySkimmer) [reply tactics/declare.ml]: Then the force can move upwards at that time, but let's not duplicate it for now. Doing useless things is confusing for the reader.

Comment (ppedrot) [reply tactics/declare.ml]: Done.

**Comment by JasonGross** [comment]
> FTR, I already have a krazy implementation of CEP 40 atop of this PR that removes the opaque table from the kernel, but does not reintroduce it anywhere else.

This is really neat!  I think it would be useful to have a flag you could pass to coqc to drop opaque proofs entirely, for decreased memory consumption at the cost of not being able to Print or Extract them

**Comment by ppedrot** [comment]
@JasonGross the goal is to store them on disk whatsoever, which should have the same effects memory-wise. As a bonus, we would also never hashcons opaque proofs to be sent to disk, for great profit performance-wise.

**Comment by JasonGross** [comment]
@ppedrot do you mean that the goal is to not store them on disk?  Or that it is to store them on disk really early, so we can free the memory at proof closing time?

**Comment by ppedrot** [comment]
@JasonGross storing them as soon as they've been type-checked, which means immediately in non-interactive mode.

File: tactics/ind_tables.ml [hunk 4]
```diff
@@ -124,8 +124,17 @@ let define internal role id c poly univs =
   let ctx = UState.minimize univs in
   let c = UnivSubst.nf_evars_and_universes_opt_subst (fun _ -> None) (UState.subst ctx) c in
   let univs = UState.univ_entry ~poly ctx in
-  let entry = Declare.definition_entry ~univs c in
-  let kn, eff = Declare.declare_private_constant ~role ~kind:Decls.(IsDefinition Scheme) ~name:id (Declare.DefinitionEntry entry) in
+  let entry = {
```
Comment (ejgallego) [inline tactics/ind_tables.ml]: Why this change? Seems spurious.

File: tactics/declare.ml (hunk 2 above)
Comment (ejgallego) [reply tactics/declare.ml]: Actually if we get the 2 PRs that make `proof_entry` totally abstract you wouldn't have more freedom to do these kind of changes.

File: tactics/ind_tables.ml (hunk 4 above)
Comment (ppedrot) [reply tactics/ind_tables.ml]: `definition_entry` has not the right type. I could export another function doing the same, though. WDYT?

Comment (ejgallego) [reply tactics/ind_tables.ml]: Likely that function already exists on my tree, let me see.

Comment (ejgallego) [reply tactics/ind_tables.ml]: Oh indeed I don't have it, I was waiting to see if `Evd.side_effects` would get a type parameter to actually match it. You can leave the code as-is if you want, and I'll introduce a `pure_definition_entry` in #10681 .

**Review by ejgallego — APPROVED** [review]
LGTM.

**Comment by gares** [comment]
>  I think it would be useful to have a flag you could pass to coqc to drop opaque proofs entirely

I think that once CEP 40 is completely implemented we can consider this request and make the whole spectrum more uniform (command line option wise):
- we can ignore opaque proofs (this gives .vos) 
- we can save the "thunk" to disk to run it later (that is .vio)
- we can run them but drop the result (that would be your option)
- we can run them and save their output (coqc default after CEP 40)
and, on an orthogonal axe, what "run" means:
- we can run them linearly and immediately as they come
- we can delegate them to workers (we do that for CoqIDE, and we never save them to the disk)

> where the best place to put the opaque table is before proceeding further

It is not very clear to me. I'd like to have the STM see that and a few years back I would have put it in library, but recently library has been split and there are PR/comments pushing parts of it in vernac, so I'm not 100% sure what to say. IMO the API to access the tables should be provided by the document manager, but it could just be an intermediary between the real storage and the client, not necessarily the one that holds the data. I guess we should discuss that part next time we meet.

**Comment by ppedrot** [comment]
Anyone assigning in the meantime?

File: kernel/safe_typing.ml [hunk 5]
```diff
@@ -860,37 +856,39 @@ let add_constant (type a) ~(side_effect : a effect_entry) l decl senv : (Constan
...
-        (* Upper layers enforce that there are no internal constraints *)
-        let () = assert (Univ.ContextSet.is_empty ctx) in
-        OpaqueDef b, Polymorphic auctx
-      | (Monomorphic _ | Polymorphic _), (Opaqueproof.PrivateMonomorphic _ | Opaqueproof.PrivatePolymorphic _) ->
-        assert false
+  kn, senv
+
+let add_private_constant l decl senv : (Constant.t * private_constants) * safe_environment =
+  let kn = Constant.make2 senv.modpath l in
+    let cb =
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml]: This `let` is incorrectly indented

File: kernel/term_typing.ml [hunk 6]
```diff
@@ -297,11 +256,46 @@ let build_constant_declaration env result =
...
+  let j = Typeops.infer env body in
+  let j = unzip ectx j in
+  let _ = Typeops.judge_of_cast env j DEFAULTcast tyj in
+  let c = j.uj_val in
+  let () = check_section_variables env declared tyj.utj_val body in
+  feedback_completion_typecheck feedback_id;
+  c, Opaqueproof.PrivateMonomorphic uctx
+| PolyTyCtx (env, tj, usubst, auctx, declared, feedback_id) ->
+  let ((body, ctx), side_eff) = body in
+  let body, ctx', _ = handle env body side_eff in
```
Comment (SkySkimmer) [inline kernel/term_typing.ml]: Shouldn't we skip_trusted here too? Only the polymorphic side effects are substituted.
We've never in history trusted side effects for an opaque polymorphic constant for some reason.

**Review by SkySkimmer — APPROVED** [review]
Seems OK

File: kernel/term_typing.ml (hunk 6 above)
Comment (ppedrot) [reply kernel/term_typing.ml]: Indeed, but this is unrelated to this PR.

**Comment by maximedenes** [comment]
@gares can you have a quick look at the private constant part of this patch?

**Comment by gares** [comment]
When you have a minute come to my office ant tell what I should look at exactly.

**Review by gares — APPROVED** [review]
LGTM

**Comment by ppedrot** [comment]
Mergy-mergy.
